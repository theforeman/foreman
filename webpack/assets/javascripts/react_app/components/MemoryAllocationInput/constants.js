export const GB_FORMAT = 'GB';

export const MB_FORMAT = 'MB';

export const GB_STEP = 1024;

export const MB_STEP = 256;
