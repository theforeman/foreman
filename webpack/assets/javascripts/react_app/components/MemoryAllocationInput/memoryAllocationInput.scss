@import '../../common/colors.scss';

.warning-icon {
  margin-right: 3px;
  color: $pf-gold-400;
}

.error-icon {
  margin-right: 3px;
  color: $pf-red-100;
}
