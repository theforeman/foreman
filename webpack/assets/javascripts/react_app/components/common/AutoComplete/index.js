import React, { Component } from 'react';
import Downshift from 'downshift';
import { Dropdown, MenuItem, InputGroup, FormControl, Button } from 'patternfly-react';
import PropTypes from 'prop-types';

export const renderItems = ({
  items, highlightedIndex, selectedItem, getItemProps,
}) => (
  <Dropdown.Menu
    style={{
      display: 'block',
      position: 'absolute',
      left: 'inherit',
      marginTop: 0,
      top: 'auto',
    }}
  >
    {items.map((item, index) => (
      <MenuItem
        {...getItemProps({
          index,
          item,
          active: highlightedIndex === index,
          onClick: (e) => {
            // At this point the event.defaultPrevented
            // is already set to true by react-bootstrap
            // MenuItem. We need to set it back to false
            // So downshift will execute it's own handler
            e.defaultPrevented = false;
          },
        })}
        key={item}
      >
        {item}
      </MenuItem>
    ))}
  </Dropdown.Menu>
);

class AutoComplete extends Component {
  constructor(props) {
    super(props);

    this.state = {
      inputValue: '',
    };
  }

  render() {
    const {
      onSearch, labelText, onInputUpdate, items, actionText, ...rest
    } = this.props;

    return (
      <Downshift
        onStateChange={({ inputValue }) => {
          if (typeof inputValue === 'string') {
            onInputUpdate(inputValue);
            this.setState({ inputValue });
          }
        }}
        defaultHighlightedIndex={0}
        selectedItem={this.state.inputValue}
        {...rest}
      >
        {({
          getInputProps,
          getItemProps,
          getLabelProps,
          isOpen,
          inputValue,
          highlightedIndex,
          getRootProps,
          selectedItem,
        }) => (
          <div>
            {labelText && <label {...getLabelProps()}>{labelText}</label>}
            <InputGroup>
              <FormControl type="text" {...getInputProps()} />
              <InputGroup.Button>
                <Button onClick={() => onSearch(inputValue)}>{actionText}</Button>
              </InputGroup.Button>
            </InputGroup>

            {isOpen && items.length
              ? renderItems({
                  items,
                  highlightedIndex,
                  selectedItem,
                  getItemProps,
                })
              : null}
          </div>
        )}
      </Downshift>
    );
  }
}

AutoComplete.propTypes = {
  items: PropTypes.arrayOf(PropTypes.string).isRequired,
  onInputUpdate: PropTypes.func.isRequired,
  onSearch: PropTypes.func.isRequired,
  labelText: PropTypes.string,
  actionText: PropTypes.string,
};

AutoComplete.defaultProps = {
  labelText: '',
  actionText: 'Search',
};

export default AutoComplete;
