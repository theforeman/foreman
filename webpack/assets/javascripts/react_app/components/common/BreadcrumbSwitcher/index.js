import React from 'react';
import {
  Popover,
  Button,
  Icon,
  ListGroup,
  ListGroupItem,
  OverlayTrigger,
} from 'patternfly-react';
import { ButtonToolbar } from 'react-bootstrap';
import EllipsisWithTooltip from 'react-ellipsis-with-tooltip';
import PropTypes from 'prop-types';
import './BreadcrumbSwitcher.scss';

const BreadcrumbSwitcher = ({ resources }) => {
  const popover = (
    <Popover id="popover-contained">
      <ListGroup className="scrollable-list">
        {resources.map((item) => {
          const { onClick, url, caption } = item;
          return (
            <ListGroupItem
              className="no-border"
              key={`id-${caption}`}
              href={url}
              onClick={onClick}
            >
              <EllipsisWithTooltip>{caption}</EllipsisWithTooltip>
            </ListGroupItem>
          );
        })}
      </ListGroup>
    </Popover>
  );
  return (
    <ButtonToolbar>
      <OverlayTrigger
        rootClose
        trigger="click"
        placement="bottom"
        overlay={popover}
      >
        <Button>
          <Icon type="fa" name="exchange" />
        </Button>
      </OverlayTrigger>
    </ButtonToolbar>
  );
};

BreadcrumbSwitcher.propTypes = {
  href: PropTypes.string,
  onClick: PropTypes.func,
  caption: PropTypes.string,
};

export default BreadcrumbSwitcher;
