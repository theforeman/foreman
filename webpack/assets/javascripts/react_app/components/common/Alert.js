import React from 'react';
import Icon from './Icon';
import { ALERT_CSS } from '../../constants';

const Alert = (props) => {
  let classNames = ALERT_CSS[props.type];

  if (props.css) {
    classNames += ' ' + props.css;
  }

  if (props.dismissable) {
    classNames += ' ' + ALERT_CSS.dismissable;
  }

  const closeBtn = (
    <button className="close" data-dismiss="alert" aria-hidden="true">
      <Icon type="close" />
    </button>
  );

  return (
    <div className={classNames}>
      {props.dismissable && closeBtn}
      {props.children}
    </div>
  );
};

Alert.defaultProps = { dismissable: false, css: null, type: 'info' };
export default Alert;
