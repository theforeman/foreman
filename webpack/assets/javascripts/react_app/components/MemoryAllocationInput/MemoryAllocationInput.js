import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { WarningTriangleIcon, ErrorCircleOIcon } from '@patternfly/react-icons';
import { HelpBlock, Grid, Col, Row } from 'patternfly-react';
import { translate as __ } from '../../common/I18n';
import NumericInput from '../common/forms/NumericInput';
import { GB_FORMAT, MB_FORMAT, GB_STEP, MB_STEP } from './constants';
import './memoryAllocationInput.scss';
import { noop } from '../../common/helpers';

const MemoryAllocationInput = ({
  label,
  defaultValue,
  onChange,
  maxValue,
  recommendedMaxValue,
  name,
  disabled,
}) => {
  const [value, setValue] = useState(defaultValue);
  const [validationState, setValidationState] = useState(undefined);

  useEffect(() => {
    if (value > maxValue && maxValue !== undefined) {
      setValidationState('error');
    } else if (
      value > recommendedMaxValue &&
      recommendedMaxValue !== undefined
    ) {
      setValidationState('warning');
    } else {
      setValidationState(undefined);
    }
  }, [recommendedMaxValue, maxValue, value]);

  const handleValueIncrease = () => {
    if (value >= GB_STEP) {
      setValue(value + GB_STEP);
    } else {
      setValue(value + MB_STEP);
    }
  };

  const handleValueDecrease = () => {
    if (value <= GB_STEP) {
      setValue(value - MB_STEP);
    } else {
      setValue(value - GB_STEP);
    }
  };

  const handleTypedValue = v => {
    if (v > GB_STEP) {
      setValue(Math.round(v / GB_STEP) * GB_STEP);
    } else {
      setValue(Math.round(v / MB_STEP) * MB_STEP);
    }
  };

  const handleChange = v => {
    if (v === value + 1) {
      handleValueIncrease();
    } else if (v === value - 1) {
      handleValueDecrease();
    } else {
      handleTypedValue(v);
    }
    onChange(value);
  };

  const format = v => {
    // used the buttons
    if (v % MB_STEP === 0) {
      if (v >= GB_STEP) {
        return `${v / GB_STEP} ${GB_FORMAT}`;
      }
      return `${v} ${MB_FORMAT}`;
    }

    // typed value
    return v;
  };

  const helpBlock = () => {
    if (validationState === 'warning') {
      return (
        <HelpBlock>
          <WarningTriangleIcon className="warning-icon" />
          {__('Specified value is higher than recommended maximum')}
        </HelpBlock>
      );
    } else if (validationState === 'error') {
      return (
        <HelpBlock>
          <ErrorCircleOIcon className="error-icon" />
          {__('Specified value is higher than maximum value')}
        </HelpBlock>
      );
    }
    return undefined;
  };

  const parser = str => str.replace(/\D/g, '');
  return (
    <Grid>
      <Row>
        <Col>
          <NumericInput
            value={value}
            format={format}
            parser={parser}
            onChange={handleChange}
            label={label}
            name={name}
            disabled={disabled}
          />
        </Col>
      </Row>
      <Row>
        <Col md={2} />
        <Col md={4} className="form-group">
          {validationState !== undefined && helpBlock()}
        </Col>
      </Row>
    </Grid>
  );
};

MemoryAllocationInput.propTypes = {
  /** Set the label of the memory allocation input */
  label: PropTypes.string,
  /** Set the default value of the numeric input */
  defaultValue: PropTypes.number,
  /** Set the recommended max value of the numeric input */
  recommendedMaxValue: PropTypes.number,
  /** Set the max value of the numeric input */
  maxValue: PropTypes.number,
  /** Set the onChange function of the numeric input */
  onChange: PropTypes.func,
  /** Set the name of the numeric input */
  name: NumericInput.propTypes.name,
  /** Set whether the numeric input will be disabled or not */
  disabled: NumericInput.propTypes.disabled,
};

MemoryAllocationInput.defaultProps = {
  label: __('Memory'),
  defaultValue: 1,
  onChange: noop,
  recommendedMaxValue: undefined,
  maxValue: undefined,
  name: '',
  disabled: false,
};

export default MemoryAllocationInput;
