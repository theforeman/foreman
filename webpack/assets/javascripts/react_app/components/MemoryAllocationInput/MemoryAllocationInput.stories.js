import React from 'react';
import { number, action, text } from '@theforeman/stories';
import MemoryAllocationInput from './MemoryAllocationInput';

export default {
  title: 'Components|Form/MemoryAllocationInput',
  component: MemoryAllocationInput,
  parameters: {
    centered: { disable: true },
  },
};

export const UseMemoryAllocationInput = () => (
  <MemoryAllocationInput
    label={text('Label', 'Memory')}
    defaultValue={number('DefaultValue', 1024)}
    onChange={action('Value was changed')}
    recommendedMaxValue={number('RecommendedMaxValue', 10240)}
    maxValue={number('MaxValue', 20480)}
  />
);
