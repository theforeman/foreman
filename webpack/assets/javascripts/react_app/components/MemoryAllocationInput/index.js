import MemoryAllocationInput from './MemoryAllocationInput';

export default MemoryAllocationInput;
