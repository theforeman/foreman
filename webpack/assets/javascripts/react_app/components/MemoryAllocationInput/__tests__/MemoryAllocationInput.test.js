import React from 'react';
import { mount } from '@theforeman/test';
import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';
import MemoryAllocationInput from '../MemoryAllocationInput';

const props = {
  label: 'Memory',
};

const fixtures = {
  'should render with default props': props,
};

describe('MemoryAllocationInput', () => {
  describe('rendering', () => {
    testComponentSnapshotsWithFixtures(MemoryAllocationInput, fixtures);
  });

  it('MB to GB change', async () => {
    const component = mount(<MemoryAllocationInput defaultValue={768} />);
    expect(component.find('input').prop('value')).toEqual('768 MB');
    component
      .find('.foreman-numeric-input-handler-up')
      .at(0)
      .simulate('mousedown');

    component.update();
    expect(component.find('input').prop('value')).toEqual('1 GB');
  });
});

it('warning alert', async () => {
  const component = mount(
    <MemoryAllocationInput defaultValue={11264} recommendedMaxValue={10240} />
  );
  expect(component.find('input').prop('value')).toEqual('11 GB');
  expect(component.find('.warning-icon').exists()).toBeTruthy();
});

it('error alert', async () => {
  const component = mount(
    <MemoryAllocationInput defaultValue={21504} maxValue={20480} />
  );
  expect(component.find('input').prop('value')).toEqual('21 GB');
  expect(component.find('.error-icon').exists()).toBeTruthy();
});
