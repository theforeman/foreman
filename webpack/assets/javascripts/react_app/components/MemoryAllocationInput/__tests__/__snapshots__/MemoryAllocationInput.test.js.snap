// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemoryAllocationInput rendering should render with default props 1`] = `
<Grid
  bsClass="container"
  componentClass="div"
  fluid={false}
>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
    >
      <NumericInput
        className=""
        disabled={false}
        format={[Function]}
        label="Memory"
        minValue={0}
        name=""
        onChange={[Function]}
        parser={[Function]}
        precision={0}
        readOnly={false}
        step={1}
        value={1}
      />
    </Col>
  </Row>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={2}
    />
    <Col
      bsClass="col"
      className="form-group"
      componentClass="div"
      md={4}
    />
  </Row>
</Grid>
`;
